docs/design/secrets.md

```diff
@@ -0,0 +1,187 @@
+3.  How will pods know when secret values change?  Should there be a way to express that a pod
+    which consumes a secret should be restarted when the secret changes?
+
+## Proposed Design
+
+### Overview
+
+This design proposes mounting secrets with a new volume type. The secrets volume will be backed
+by a volume plugin that does the actual work of fetching the secret and placing it on the
+filesystem to be mounted in the container. It should be possible to mount single files as part of a
+secret. Secrets may consist of multiple files. For example, an SSH key pair. In order to remove the
+burden from the end user in specifying every file that a secret consists of, it should be possible
+to mount all files provided by a secret with a single ```VolumeMount``` entry in the container
+specification.
+
+### Secret Volume Source
+
+A new Secret type of volume source will be added to the ```VolumeSource``` struct in the API:
+
+```go
+type VolumeSource struct {
+     ... 
+     Secret *Secret `json:"secret"`
+}
+
+type Secret struct {
+     // Ref is a reference to the actual secret
+     Ref string 
+     
+     // Files is a set of descriptors of where and how the secret files
+     // should be placed (optional). If not specified, the secret itself
+     // will determine where its files should go
+	 Files []SecretFile
+}
+
+type SecretFile struct {
+     // Name of the file within the secret. For example, in the case
+     // of SSH keys, you would have a 'private' and a 'public' key
+	 Name string
+	 
+	 // Path of the secret within the container
+	 Path string
+	 
+	 // Owner id of the file owner
+	 OwnerUID int
+	 
+	 // Group id of the file group
+	 GroupUID int
+	 
+	 // File Mode
+	 Mode os.FileMode
+	 
+	 // MCSLabel for SELinux
+	 MCSLabel string
+}
+
+```
+
+### Secret Volume Plugin
+
+The secret volume plugin would implement the actual retrieval and laying down of secrets on the
+Kubelet's file system. Secrets may be stored in a special secret registry, as Docker volumes, LDAP
+etc. See [Issue #2030](https://github.com/GoogleCloudPlatform/kubernetes/issues/2030) for options.
+The implementation of this plugin is outside of the scope of this proposal. A default Etcd-based
+version could be provided out of the box, but different solutions may be appropriate based on the
+use case for Kubernetes.
+
+### Changes to Support Secret Volumes
+
+Because secrets require mounting multiple files, the way bind mounts are generated from pods would
+need to be modified to allow the pod spec to say "I want the default mountpoints" instead of a
+specific MountPath. This can be done by modifying the VolumeMount struct to include a boolean:
+
+```go
+type VolumeMount struct {
+	Name string
+	ReadOnly bool
+	MountPath string // Change from required to optional
+	
+	// New boolean to let the Kubelet know we want the default path
+	// from the volume
+	DefaultPath bool 
```

Comment:
Yup, an example might help. But basically it is assumed that there is a contract between the container and the secret about the location of the files. If they are stored as kube resources, this would be part of the kube resource definition.